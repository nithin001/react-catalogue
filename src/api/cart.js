import instance from './base';
import Ajv from 'ajv';

export const get = () => {
  return new Promise((resolve, reject) => {
    instance.get('/cart')
      .then((response) => {
        if (response.data === undefined) {
          reject('backend_error');
          return;
        }
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(get_schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        reject('backend_error');
      });
  });
};

export const put = (request) => {
  return new Promise((resolve, reject) => {
    instance.put('/cart', { data: request })
      .then((response) => {
        if (response.data === undefined) {
          reject('backend_error');
          return;
        }
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(put_schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        reject('backend_error');
      });
  });
};

const get_schema = {
  '$id': 'http://challenge.monoqi.net/cart',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'properties': {
    'lines': {
      '$id': '/properties/lines',
      'type': 'array',
      'items': {
        '$id': '/properties/lines/items',
        'type': 'object',
        'properties': {
          'sku': {
            '$id': '/properties/lines/items/properties/sku',
            'type': 'string',
            'title': 'The Sku Schema ',
            'default': '',
            'examples': [
              '199203'
            ]
          },
          'quantity': {
            '$id': '/properties/lines/items/properties/quantity',
            'type': 'integer',
            'title': 'The Quantity Schema ',
            'default': 0,
            'examples': [
              3
            ]
          }
        },
        'required': ['sku', 'quantity']
      }
    }
  },
  'required': ['lines']
};

const put_schema = {
  '$id': 'http://example.com/example.json',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'properties': {
    'lines': {
      '$id': '/properties/lines',
      'type': 'array',
      'items': {
        '$id': '/properties/lines/items',
        'type': 'object',
        'properties': {
          'sku': {
            '$id': '/properties/lines/items/properties/sku',
            'type': 'string',
            'title': 'The Sku Schema ',
            'default': '',
            'examples': [
              '199203'
            ]
          },
          'name': {
            '$id': '/properties/lines/items/properties/name',
            'type': 'string',
            'title': 'The Name Schema ',
            'default': '',
            'examples': [
              'IPad'
            ]
          },
          'price': {
            '$id': '/properties/lines/items/properties/price',
            'type': 'object',
            'properties': {
              'amount': {
                '$id': '/properties/lines/items/properties/price/properties/amount',
                'type': 'integer',
                'title': 'The Amount Schema ',
                'default': 0,
                'examples': [
                  229
                ]
              },
              'currency': {
                '$id': '/properties/lines/items/properties/price/properties/currency',
                'type': 'string',
                'title': 'The Currency Schema ',
                'default': '',
                'examples': [
                  'EUR'
                ]
              }
            },
            'required': [
              'amount',
              'currency'
            ]
          },
          'quantity': {
            '$id': '/properties/lines/items/properties/quantity',
            'type': 'integer',
            'title': 'The Quantity Schema ',
            'default': 0,
            'examples': [
              3
            ]
          },
          'lineTotal': {
            '$id': '/properties/lines/items/properties/lineTotal',
            'type': 'object',
            'properties': {
              'amount': {
                '$id': '/properties/lines/items/properties/lineTotal/properties/amount',
                'type': 'integer',
                'title': 'The Amount Schema ',
                'default': 0,
                'examples': [
                  687
                ]
              },
              'currency': {
                '$id': '/properties/lines/items/properties/lineTotal/properties/currency',
                'type': 'string',
                'title': 'The Currency Schema ',
                'default': '',
                'examples': [
                  'EUR'
                ]
              }
            },
            'required': [
              'amount',
              'currency'
            ]
          }
        },
        'required': [
          'sku',
          'name',
          'price',
          'quantity',
          'lineTotal'
        ]
      }
    },
    'total': {
      '$id': '/properties/total',
      'type': 'object',
      'properties': {
        'amount': {
          '$id': '/properties/total/properties/amount',
          'type': 'integer',
          'title': 'The Amount Schema ',
          'default': 0,
          'examples': [
            1126
          ]
        },
        'currency': {
          '$id': '/properties/total/properties/currency',
          'type': 'string',
          'title': 'The Currency Schema ',
          'default': '',
          'examples': [
            'EUR'
          ]
        }
      },
      'required': [
        'amount',
        'currency'
      ]
    }
  },
  'required': [
    'lines',
    'total'
  ]
};