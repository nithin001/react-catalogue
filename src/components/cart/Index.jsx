import React from 'react';
import { connect } from 'react-redux';
import CartItem from './CartItem';
export const Cart = ({ cart }) => {
  const skus = [];
  cart.keySeq().forEach(key => skus.push(key));
  const items = skus.map((key) => (<CartItem
    key={key}
    sku={key}
    quantity={cart.get(key)}
  />));
  return <div className={'cart'}>{items}</div>;
};

export const mapStateToProps = (state) => {
  return { cart: state.cart };
};

export default connect(mapStateToProps, null)(Cart);
