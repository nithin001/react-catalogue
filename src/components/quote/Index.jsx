import React from 'react';
import { connect } from 'react-redux';
import { Total } from './Total';
import { QuoteItem } from './QuoteItem';
export const Quote = ({ quote, loading }) => {

  if (quote.get('lines') === undefined || quote.get('lines').size == 0) {
    return null;
  }

  const quoteItems = quote.get('lines').map(quoteItem => {
    return (<QuoteItem key={quoteItem.get('sku')} order={quoteItem}/>);
  });

  return (
    <div>
      <div className={'quote-items'}>
        <div class="card blue-grey darken-1">
          <div class="card-content white-text">
            {quoteItems}
          </div>
          <div class="card-action white-text">
            <Total total={quote.get('total')}/>
          </div>
        </div>

      </div>

    </div>
  );
};

export const mapStateToProps = (state) => {
  return { quote: state.quote };
};

export default connect(mapStateToProps, null)(Quote);
