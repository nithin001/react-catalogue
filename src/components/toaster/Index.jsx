import React from 'react';
import { connect } from 'react-redux';
import materialize from 'materialize-css';
export const Toaster = ({ message }) => {
  if (message !== '') {
    materialize.toast(message, 1000);
  }
  return null;
};

export const mapStateToProps = (state) => {

  let message = '';
  if (state.ui.get('cart_loading')) {
    message = 'Loading cart';
  } else if (state.ui.get('quote_loading')) {
    message = 'Loading quotation';
  } else if (state.ui.get('article_loading')) {
    message = 'Loading article';
  } else if (state.ui.get('catalog_loading')) {
    message = 'Loading catalog';
  } else if (state.ui.get('error')) {
    message = 'There was an error. Please try again later!';
  }
  return { message };
};

export default connect(mapStateToProps, null)(Toaster);


