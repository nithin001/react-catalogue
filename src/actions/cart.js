import { get } from '../api/cart';
export const load = () => {
  return (dispatch) => {
    dispatch({ type: 'LOAD_CART_START' });
    return get()
      .then((cart) => {
        dispatch({ type: 'POPULATE_CART', payload: cart });
        dispatch({ type: 'LOAD_CART_COMPLETE' });
      })
      .catch(() => {
        dispatch({ type: 'LOAD_ERROR' });
      });
  };
};