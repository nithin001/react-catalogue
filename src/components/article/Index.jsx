import React from 'react';
import { connect } from 'react-redux';
import { addToCart } from '../../actions/cart';

export const Article = ({ article, addToCart }) => {
  return <div className={'article'}>
    <span className={'name'}>{article.get('name')}</span>
    <span className={'description'}>{article.get('description')}</span>
    <img src={article.get('image')} className={'image'}/>
    <span className={'price-amount'}>{article.getIn(['price', 'amount'])}</span>
    <span className={'price-currency'}>{article.getIn(['price', 'currency'])}</span>
    <button className={'add-to-cart'} onClick={() => {
      addToCart(article.get('sku'));
    }}>Add to cart
    </button>
  </div>;
};

export const mapStateToProps = (state) => {
  return { article: state.article };
};

export const mapDispatchToProp = (dispatch) => ({
    addToCart: (sku) => dispatch(addToCart(sku)),
  }
);

export default connect(mapStateToProps, mapDispatchToProp)(Article);
