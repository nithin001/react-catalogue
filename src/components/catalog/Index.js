import React from 'react';
import { connect } from 'react-redux';
import { CatalogItem } from './CatalogItem';
import { addToCart } from '../../actions/cart';
export const Catalog = ({ catalog, addToCart }) => {
  const articles = [];
  catalog.keySeq().forEach((item) => {
    articles.push(catalog.get(item));
  });
  const articlesList = articles.map(item => {
    return (<CatalogItem key={item.get('sku')} item={item} addToCart={addToCart}/>);
  });
  return (<div className={'catalog'}>
    {articlesList}
  </div>);
};

export const mapStateToProps = (state, ownProps) => {
  return { catalog: state.catalog };
};

export const mapDispatchToProp = (dispatch) => ({
    addToCart: (sku) => dispatch(addToCart(sku)),
  }
);

export default connect(mapStateToProps, mapDispatchToProp)(Catalog);


