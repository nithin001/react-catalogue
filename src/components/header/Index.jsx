import React from 'react';
import { connect } from 'react-redux';

import Title from './Title';
import CartTeaser from './CartTeaser';
import { load as loadCart } from '../../actions/cart';
import { load as loadQuote } from '../../actions/quote';
export class Header extends React.Component {
  constructor (props) {
    super(props);
    if (!this.props.cartLoaded) {
      this.props.loadCart();
    }
  }

  componentDidUpdate () {
    const cartItems = [];
    this.props.cart.forEach((value, key) => {
        cartItems.push(key);
      }
    );
    const cart = cartItems.map(item => ({
      sku: item,
      quantity: this.props.cart.get(item)
    }));
    this.props.loadQuote(cart);
  }

  render () {
    return (<div className={'navbar-fixed'}>
      <nav >
        <div className={'nav-wrapper'}>
          <Title/>
          <CartTeaser/>
        </div>
      </nav>
    </div>);
  }
}

export const mapStateToProps = (state) => {
  const cartLoaded = state.ui.get('cart_loaded_from_server');
  return { cartLoaded, cart: state.cart };
};

export const mapDispatchToProps = (dispatch) => {
  return {
    loadQuote: (cart) => dispatch(loadQuote(cart)),
    loadCart: () => dispatch(loadCart())
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(Header);
