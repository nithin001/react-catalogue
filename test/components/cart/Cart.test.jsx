import React from 'react';
import { configure, shallow } from 'enzyme';
import Immutable from 'immutable';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { Cart, mapStateToProps } from '../../../src/components/cart/Index';
import cart from '../../../src/reducers/cart';
configure({ adapter: new Adapter() });
var mockCart = {
  lines: [{
    'sku': '199203',
    'quantity': 3
  }, {
    'sku': '199204',
    'quantity': 5
  }]
};
const cartState = cart(undefined, { type: 'POPULATE_CART', payload: mockCart });
describe('Cart component', () => {
  it('should render the list of CartItem components', () => {
    const props = { cart: cartState };
    const cartObj = shallow(<Cart
      loading={false}
      {...props}
    />);
    expect(shallowToJson(cartObj)).toMatchSnapshot();
  });

  it('should render loading if loading', () => {
    const props = { cart: cartState };
    const cartObj = shallow(<Cart
      loading={true}
      {...props}
    />);
    expect(shallowToJson(cartObj)).toMatchSnapshot();
  });

  it('should render empty cart', () => {
    const props = { cart: Immutable.Map() };
    const cartObj = shallow(<Cart
      loading={false}
      {...props}
    />);
    expect(shallowToJson(cartObj)).toMatchSnapshot();
  });
});

describe('Connected Cart component', () => {
  it('should map state values to props of the component', () => {
    const state = {
      cart: cartState,
    };
    const result = mapStateToProps(state);
    expect(result).toEqual({ cart: cartState });
  });
});