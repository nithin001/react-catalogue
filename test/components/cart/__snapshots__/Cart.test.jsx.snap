// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component should render the list of CartItem components 1`] = `
<div
  className="cart"
>
  <Connect(CartItem)
    key="199203"
    quantity={3}
    sku="199203"
  />
  <Connect(CartItem)
    key="199204"
    quantity={5}
    sku="199204"
  />
</div>
`;
